### ISSUE TITLE
Component template changes not reflected in app 

### ISSUE THREAD

**Issue by allmarkedup** [github/view_component#1426]
Hey all. I'm running into an issue after upgrading ViewComponent to >= v2.58.0 whereby any changes I make to component view templates are not showing up when the component is rendered in our app views.

Changes to the component class file **are** picked up correctly - and If I make a change to a template and *then* make a change to the component class file, the template change will then show up in the app correctly.

Otherwise, the only way to see the changes to the component view template is to restart the server. 

I'm pretty sure this is somehow related to our specific app setup, as a fresh install of the latest ViewComponent version in a new project works as expected. But I'm tearing my hair out a little bit here so I was wondering if anyone had any ideas as to what might be going on and help guide my debugging efforts!

Any help on this would be very much appreciated :-)

### Steps to reproduce

- Start app.
- Make change to a component view template file
- View an instance of the component rendered in the app

### Expected behavior

I should see the change to the component template reflected in the app.

### Actual behavior

The component output is unchanged from how it was before the template was edited.

### System configuration
**Rails version**:
6.1.6.1

**Ruby version**:
2.7.1

**Gem version**:
\>= v2.58.0

**Comment by joelhawksley** [comment]
@allmarkedup thanks for the report! Might you be able to bisect this problem against `main`, testing whether https://github.com/github/view_component/pull/1343 broke this for you?

**Comment by allmarkedup** [comment]
Hey @joelhawksley - I've just checked the latest release and I am still seeing the same issue, so I've just done a bit of digging and it seems that PR https://github.com/github/view_component/pull/1343 is the cause of the problem for me. Before that everything works fine.

Just skim-reading the discussion around that PR, might it be possibly something to do that all our components inherit from a base Component class that contains a `call` method? Is that likely to be an issue?

Each of our components then either provide a template, override the parent base component `call` method or just fall back to using the inherited `call` method (which basically just renders a tag with any attributes defined in the `initialize` method).

So the set up is something like this (greatly simplified!):

```ruby
class BaseComponent < ViewComponent::Base
  def initialize(**html_attrs)
    @html_attrs = html_attrs
  end

  def call
    tag.div **@html_attrs do
      content
    end
  end
end
```

```ruby
# example_component.rb
class ExampleComponent < BaseComponent
  def initialize(title: "A title", **html_attrs)
    @title = title
    super(**html_attrs)
  end
end
```

```erb
<!-- example_component.html.erb -->
<%= tag.section **@html_attrs do %>
  <h2><%= title %></h2>
  <%= content %>
<% end %>
```

**Comment by allmarkedup** [comment]
@joelhawksley I've just tried removing the `call` method from our BaseComponent and without it the template changes are once again being picked up in the app.

So it looks like it's something to do with `call` method inheritance - is this an anti-pattern? It would be pretty trivial to refactor out of our codebase but would be good to know what you'd recommend :-)

**Comment by camertron** [comment]
@allmarkedup I don't believe we consider `call` method inheritance + templates to be an anti-pattern, everything should work as expected. I just copy/pasted your components into a new Rails app and was not able to reproduce the behavior you're seeing. Is it possible fragment caching (or some other caching mechanism) is at fault here?

One thing I did notice: if you make an initial request and then _delete_ example_component.html.erb, the next request will error saying it can't find the template file. In other words, deleting the template while the app is running causes problems. The only way to hit `BaseComponent#call` was to restart the app. I'm not sure if that's related or not.

We did just remove some code in https://github.com/github/view_component/pull/1432 that should have made this sort of problem a thing of the past. Dumb question, but can you verify you're using the latest version of view_component?

I would be happy to debug this further, but I'm kinda stuck without a reproducible test case. The next step would be to get a test app to demonstrate the problem that I can clone and hack on.

**Comment by allmarkedup** [comment]
Hey @camertron - I've just tested with the latest release that includes the changes in https://github.com/github/view_component/pull/1432 and I can confirm that the issue I was seeing is now fixed. Thank you!

FWIW, we are not doing any fragment caching and it wasn't anything to do with deleting templates while the app was running. But whatever the cause it is now resolved :-) Appreciate your help with this.
